### DISCUSSION TITLE
Custom fields | Fail2ban Heatmap

### DISCUSSION THREAD

**Discussion by alafad** [wazuh/wazuh#14917] [Q&A] [answered]
Hello, i would like create an Heatmap with all banned fail2ban IPs with Geoip.

But a have only the scrip field, how can i ad an fail2banip field, the following Documentations doesnt had helped me:

- https://documentation.wazuh.com/current/user-manual/ruleset/dynamic-fields.html
- https://documentation.wazuh.com/current/user-manual/ruleset/json-decoder.html

Is it possible to create such a Heatmap with a custom Decoder or rule?

And can anybody please give me an example for my idea?

my Decoder:

```
<decoder name="fail2ban_dec">
   <prematch>[\d+]: NOTICE\s+[\S+]|[\d+]: INFO\s+[\S+]|[\d+]: WARNING\s+[\S+]|[\d+]: ERROR\s+[\S+]</prematch>
</decoder>


<decoder name="fail2ban_dec_ip">
   <parent>fail2ban_dec</parent>
   <regex>Ban (\d+.\d+.\d+.\d+)|Unban (\d+.\d+.\d+.\d+)</regex>
   <order>srcip,fail2banip</order>
</decoder>
```

my Ruleset:

```
<group name="local,syslog,sshd,fail2ban_dec,fail2ban_dec_ip,">

 <rule id="100300" level="3">
  <regex> Ban \d+.\d+.\d+.\d+</regex>
    <description>IP Banned by Fail2ban</description>
  <group>cuh_custom</group>
 </rule>

</group>
```

**Comment by s-ocando** [comment, marked as answer]
In your `fail2ban_dec_ip` decoder there's no need for a second variable. In both cases, Ban or Unban, the IP will be stored under `scrip`.
```
<decoder name="fail2ban_dec_ip">
   <parent>fail2ban_dec</parent>
   <regex>Ban (\d+.\d+.\d+.\d+)|Unban (\d+.\d+.\d+.\d+)</regex>
   <order>srcip</order>
</decoder>
```
You can verify this using the `/var/ossec/bin/wazuh-logtest` tool and a typical fail2ban log `2021-02-24 13:22:49,234 fail2ban.actions        [6975]: NOTICE  [sshd] Unban 149.202.162.73`:
```
root@master:/home/vagrant# /var/ossec/bin/wazuh-logtest
Starting wazuh-logtest v4.3.7
Type one log per line

2021-02-24 13:22:49,234 fail2ban.actions        [6975]: NOTICE  [sshd] Unban 149.202.162.73

**Phase 1: Completed pre-decoding.
	full event: '2021-02-24 13:22:49,234 fail2ban.actions        [6975]: NOTICE  [sshd] Unban 149.202.162.73'
	timestamp: '2021-02-24 13:22:49,23'

**Phase 2: Completed decoding.
	name: 'fail2ban_dec'
	srcip: '149.202.162.73'
```	
Note that the IP is decoded as `srcip`, so you may remove the `fail2banip` field from the decoder.
As `scrip` is enriched with geographical information by default, your alerts will automatically include GeoLocation.  As a reference, if you want to enrich a field that's not enriched by default, you can modify the Filebeat pipeline `/usr/share/filebeat/module/wazuh/alerts/ingest/pipeline.json` to include your field. For example:
```
    {
      "geoip": {
        "field": "data.fail2banip",
        "target_field": "Fail2BanGeoLocation",
        "properties": ["city_name", "country_name", "region_name", "location"],
        "ignore_missing": true,
        "ignore_failure": true
      }
    },
 ```    
After modifying the pipeline, it's necessary to update it so changes can take effect:  `filebeat setup --pipelines`

**Reply by alafad** [reply]
Thanks for your reply, but i dont understand , how can i differ between srcip and the fail2banip. How can i configure the decoder to defining the ip as fail2banip and not as a src.ip or create another field with failwbanip.

Because each fail2banip is a srcip nut not each srcip is a fail2banip.

For example, i would like a Output like this :

```

**Phase 2: Completed decoding.
	name: 'fail2ban_dec'
	FAIL2BANIP: '149.202.162.73'
```

**Reply by s-ocando** [reply]
Hi! If you want to create a Heatmap with all the IP addresses banned by _fail2ban_, you may do so by creating a **Coordinate Map** and filtering your _fail2ban_ rules, for example, `100300`. 

![image](https://user-images.githubusercontent.com/61882981/190658898-d515b129-6f65-4e28-bf8c-8d84f226f1db.png)

Is this what you're looking for? Or is there a particular reason why you want to name your field `fail2banip` instead of `srcip`? 
Let us know what's your use case so we can guide you on how to achieve your goal.

**Reply by alafad** [reply]
Thank you, that helps me a lot.

**Reply by s-ocando** [reply]
You're welcome @alafad, I'm glad I could help you. Please mark the thread as answered to help other users who may have the same question.
